source/language/openpulse.rst

```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
-Restrictions on the use of frames
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The frame is composed of three parts:
 
-There are two important considersations when dealing with frames.
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time of type ``dt`` which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``.
 
-First, frames are compile time constants. Runtime variables cannot be used as
-arguments to frames; the frame must be resolvable at compile time. This also
-means that assigning a frame to an alias (for the purpose of avoiding typing)
-needs to be done with ``const`` not ``let``.
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
 
-.. code-block: none
+Frame Construction
+~~~~~~~~~~~~~~~~~~
 
-   const driveframe = frameof("drive", %0);
+Frames are purely virtual and can be constructed using the ``Frame`` command
 
-Second, frames return references not values. This means that calling the same
-``frameof`` command in different places is the same as assigning the frame to a
-constant and referencing it multiple times.
+.. code-block:: javascript
 
-Frame manipulation
-~~~~~~~~~~~~~~~~~~
+  frame driveframe = Frame(5e9, 0.0);
 
-The (frequency, phase, scale) of a frame can be manipulated throughout program
-by referencing ``.frequency``, ``.phase``, and ``.scale``. Operations must be 
-appropriate for the respective type, ``float`` for frequency, ``angle`` for
-phase, and ``float`` for scale. Again, the exact precision of these calculations
-is hardware specific.
+When instantiated, the frame time starts at 0. ``Frame``s can also be copied using the
+``copy`` command with argument replacement
 
-Here's an example of manipulating the phase to calibrate an ``rz`` gate:
+.. code-block:: javascript
 
-.. code-block:: none
+  frame driveframe1 = Frame(1.0, 5e9, 0.0);
+  frame driveframe2 = copy(driveframe1, phase=driveframe1.phase + pi/2);
 
-   // Shift phase of qubit 0's "drive" frame by pi/4, eg. an rz gate with angle -pi/4
-   frameof("drive", %0).phase += pi/4;
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` (i.e. with
+the same frequency and time as ``driveframe1``) with the same time as `driveframe1`.
 
-   // Equivalent
-   const drive = frameof("drive", %0);
-   drive.phase += pi/4;
+Like other things in OpenQASM, a ``frame`` is locally scoped unless a ``prelude`` defcal is used
+to instantiate frames in global scope.
 
-   // Define a calibration for the rz gate on all physical qubits
-   defcal rz(angle[20]:theta) %q {
-     frameof("drive", %q).phase -= theta;
+.. code-block:: javascript
+
+   defcal prelude {
```

Comment:
I was wondering if we should consider getting qubit wildcarded attributes, this would allow more generic defcals that work for many qubits, eg., 

```c
defcal prelude {
     set_attr("frame",  Frame(5.0e9, 0)) $0;
}

defcal rz(angle[20] theta) $q {
     frame carrier = get_attr("frame") $q;
     carrier.phase -= theta;
}